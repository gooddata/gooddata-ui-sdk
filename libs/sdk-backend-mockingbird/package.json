{
    "name": "@gooddata/sdk-backend-mockingbird",
    "version": "8.6.0-alpha.41",
    "author": "GoodData",
    "description": "Mock GoodData Backend SPI implementation",
    "repository": {
        "type": "git",
        "url": "https://github.com/gooddata/gooddata-ui-sdk.git",
        "directory": "libs/sdk-backend-mockingbird"
    },
    "license": "MIT",
    "main": "dist/index.js",
    "module": "esm/index.js",
    "browser": "dist/index.js",
    "typings": "esm/index.d.ts",
    "sideEffects": false,
    "files": [
        "dist/**/*.js",
        "dist/**/*.json",
        "dist/**/*.d.ts",
        "dist/**/*.map",
        "esm/**/*.js",
        "esm/**/*.json",
        "esm/**/*.d.ts",
        "esm/**/*.map"
    ],
    "config": {
        "eslint": "-c .eslintrc.js --ext ts src/"
    },
    "scripts": {
        "clean": "rm -rf ci dist esm coverage *.log && jest --clearCache",
        "dev": "tsc -p tsconfig.dev.json --watch",
        "build": "concurrently \"npm run build-cjs\" \"npm run build-esm\" && npm run api-extractor",
        "build-cjs": "tsc -p tsconfig.build.json",
        "build-esm": "tsc -p tsconfig.build.esm.json",
        "api-extractor": "mkdir -p api && [ -z \"${CI}\" ] && (api-extractor run -l) || (api-extractor run)",
        "test": "jest --watch",
        "test-once": "jest",
        "test-ci": "JEST_JUNIT_OUTPUT=./ci/results/test-results.xml jest --ci --config jest.ci.js",
        "eslint": "eslint $npm_package_config_eslint",
        "eslint-ci": "mkdir -p ./ci/results && eslint -f checkstyle -o ci/results/eslint-results.xml $npm_package_config_eslint",
        "prettier-check": "prettier --check '{src,test}/**/*.{ts,tsx,json,scss,md,yaml,html}'",
        "prettier-write": "prettier --write '{src,test}/**/*.{ts,tsx,json,scss,md,yaml,html}'",
        "dep-cruiser": "depcruise --validate .dependency-cruiser.js --output-type err-long src/",
        "dep-cruiser-ci": "depcruise --validate .dependency-cruiser.js --output-type err-long src/",
        "validate": "npm run dep-cruiser && npm run eslint && npm run prettier-check",
        "validate-ci": "npm run dep-cruiser-ci && npm run eslint-ci && npm run prettier-check"
    },
    "dependencies": {
        "@gooddata/sdk-backend-base": "^8.6.0-alpha.41",
        "@gooddata/sdk-backend-spi": "^8.6.0-alpha.41",
        "@gooddata/sdk-model": "^8.6.0-alpha.41",
        "lodash": "^4.17.19",
        "ts-invariant": "^0.7.3",
        "tslib": "^2.0.0"
    },
    "devDependencies": {
        "@gooddata/api-model-bear": "^8.6.0-alpha.41",
        "@gooddata/eslint-config": "^2.1.0",
        "@gooddata/reference-workspace": "^8.6.0-alpha.41",
        "@microsoft/api-documenter": "^7.12.13",
        "@microsoft/api-extractor": "^7.18.4",
        "@types/jest": "^26.0.12",
        "@types/lodash": "^4.14.158",
        "@typescript-eslint/eslint-plugin": "^4.14.2",
        "@typescript-eslint/parser": "^4.14.2",
        "concurrently": "^6.0.2",
        "dependency-cruiser": "^9.9.5",
        "eslint": "^7.5.0",
        "eslint-plugin-header": "^3.0.0",
        "eslint-plugin-import": "^2.22.0",
        "eslint-plugin-jest": "^24.1.3",
        "eslint-plugin-no-only-tests": "^2.4.0",
        "eslint-plugin-prettier": "^3.1.4",
        "eslint-plugin-sonarjs": "^0.9.1",
        "jest": "^26.4.2",
        "jest-junit": "^3.0.0",
        "prettier": "~2.3.0",
        "ts-jest": "^26.3.0",
        "typescript": "4.0.2"
    }
}
