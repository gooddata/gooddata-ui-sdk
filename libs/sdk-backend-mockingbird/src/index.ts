// (C) 2019-2024 GoodData Corporation
/**
 * This package provides a mock Analytical Backend implementation used mainly for testing.
 *
 * @remarks
 * For the GoodData Cloud and GoodData.CN version, see `@gooddata/sdk-backend-tiger`.
 *
 * @packageDocumentation
 */
import { dummyBackend, dummyBackendEmptyData, dummyDataView } from "@gooddata/sdk-backend-base";
export { dummyBackend, dummyBackendEmptyData, dummyDataView };

export { recordedBackend, defaultRecordedBackendCapabilities } from "./recordedBackend/index.js";

export type {
    RecordingIndex,
    ExecutionRecording,
    InsightRecording,
    DisplayFormRecording,
    ScenarioRecording,
    RecordedBackendConfig,
    CatalogRecording,
    VisClassesRecording,
    DashboardRecording,
    RecordedRefType,
    SecuritySettingsUrlValidator,
    SecuritySettingsOrganizationScope,
    SecuritySettingsPluginUrlValidator,
    IUserGroup,
    IUsers,
    IAccessControl,
    IUserManagement,
    AttributeElementsFiltering,
    AttributeElementsFilteringPredicate,
} from "./recordedBackend/types.js";

export type { NamedDataView } from "./recordedBackend/execution.js";
export {
    recordedDataView,
    recordedDataViews,
    recordedInsight,
    recordedInsights,
    DataViewAll,
    dataViewWindow,
    DataViewFirstPage,
} from "./recordedBackend/execution.js";

export {
    newAttributeFilterLimitingItem,
    newDateFilterLimitingItem,
    newMeasureLimitingItem,
} from "./recordedBackend/elementsUtils.js";

export type { CompositeBackendPart } from "./compositeBackend/index.js";
export { compositeBackend } from "./compositeBackend/index.js";

export type {
    LegacyExecutionRecording,
    LegacyRecordingIndex,
    LegacyWorkspaceRecordings,
} from "./legacyRecordedBackend/index.js";
export { legacyRecordedBackend, legacyRecordedDataView } from "./legacyRecordedBackend/index.js";

export { objRefsToStringKey } from "./recordedBackend/utils.js";
