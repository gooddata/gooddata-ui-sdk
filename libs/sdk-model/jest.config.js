// (C) 2019 GoodData Corporation
import base from "../../common/config/jest/jest.config.base.mjs";

export default {
  ...base,
};