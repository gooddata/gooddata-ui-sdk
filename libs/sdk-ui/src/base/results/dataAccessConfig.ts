// (C) 2019-2023 GoodData Corporation
import { DataValue, ISeparators } from "@gooddata/sdk-model";
import { ClientFormatterFacade } from "@gooddata/number-formatter";
import escape from "lodash/escape.js";
import unescape from "lodash/unescape.js";

const customEscape = (str: string) => str && escape(unescape(str));

/**
 * @public
 */
export type ValueFormatter = (value: DataValue, format: string) => string;

/**
 * @public
 */
export type HeaderTranslator = (value: string | null) => string;

/**
 * Creates value formatter that uses `@gooddata/number-formatter` to format raw measure values according
 * to the format string.
 *
 * @remarks
 * By default, the format will strip away all the coloring information and
 * just return the value as string.
 *
 * @param separators - number separators to use. if not specified then `numberjs` defaults will be used
 * @public
 */
export function createNumberJsFormatter(separators?: ISeparators): ValueFormatter {
    return (value: DataValue, format: string) => {
        const valueToFormat = ClientFormatterFacade.convertValue(value);
        const { formattedValue } = ClientFormatterFacade.formatValue(valueToFormat, format, separators);
        return customEscape(formattedValue);
    };
}

/**
 * Default configuration for the data access methods. Uses default `@gooddata/number-formatter` formatter and no result formatting.
 *
 * @public
 */
export const DefaultDataAccessConfig: DataAccessConfig = {
    valueFormatter: createNumberJsFormatter(),
};

/**
 * @public
 */
export type DataAccessConfig = {
    /**
     * Function to use to format measure values.
     */
    valueFormatter: ValueFormatter;

    /**
     * Function to translate header names.
     */
    headerTranslator?: HeaderTranslator;
};
