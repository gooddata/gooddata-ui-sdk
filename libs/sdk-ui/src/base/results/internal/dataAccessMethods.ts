// (C) 2019-2022 GoodData Corporation

import {
    IDataAccessMethods,
    IDataSeries,
    IDataSeriesCollection,
    IDataSlice,
    IDataSliceCollection,
} from "../dataAccess.js";
import { IDataView } from "@gooddata/sdk-backend-spi";
import { DataAccessImpl } from "./dataAccessImpl.js";
import { DataAccessConfig, DefaultDataAccessConfig } from "../dataAccessConfig.js";
import {
    IMeasure,
    IAttribute,
    ITotal,
    measureLocalId,
    IMeasureDescriptor,
    IAttributeDescriptor,
} from "@gooddata/sdk-model";

class FilteredIterator<T> implements Iterator<T> {
    private idx = 0;

    constructor(private readonly indexes: number[], private readonly itemProvider: (idx: number) => T) {}

    public next = (): IteratorYieldResult<T> | IteratorReturnResult<any> => {
        if (this.idx >= this.indexes.length) {
            return {
                done: true,
                value: undefined,
            };
        }

        const value: T = this.itemProvider(this.indexes[this.idx]);
        this.idx += 1;

        return {
            done: false,
            value,
        };
    };
}

class DataSeriesCollection implements IDataSeriesCollection {
    public readonly count: number = 0;
    public readonly fromMeasures: IMeasureDescriptor[] = [];
    public readonly fromMeasuresDef: IMeasure[] = [];
    public readonly scopingAttributes: IAttributeDescriptor[] = [];
    public readonly scopingAttributesDef: IAttribute[] = [];

    constructor(private readonly dataAccess: DataAccessImpl) {
        const seriesDigest = this.dataAccess.getDataAccessPointers().series;

        if (!seriesDigest) {
            return;
        }

        this.count = seriesDigest.count;
        this.fromMeasures = seriesDigest.fromMeasures;
        this.fromMeasuresDef = seriesDigest.fromMeasuresDef;
        this.scopingAttributes = seriesDigest.scopingAttributes;
        this.scopingAttributesDef = seriesDigest.scopingAttributesDef;
    }

    public [Symbol.iterator] = (): Iterator<IDataSeries> => {
        return this.dataAccess.getDataSeriesIterator();
    };

    public allForMeasure = (localIdOrMeasure: string | IMeasure): Iterable<IDataSeries> => {
        if (!this.count) {
            return [];
        }

        const localId =
            typeof localIdOrMeasure === "string" ? localIdOrMeasure : measureLocalId(localIdOrMeasure);
        const indexes = this.dataAccess.findDataSeriesIndexes(localId);

        if (!indexes.length) {
            return [];
        }

        return {
            [Symbol.iterator]: (): Iterator<IDataSeries> => {
                return new FilteredIterator(indexes, this.dataAccess.getDataSeries);
            },
        };
    };

    public firstForMeasure = (localIdOrMeasure: string | IMeasure): IDataSeries => {
        if (!this.count) {
            throw new Error("there are no data series");
        }

        const localId =
            typeof localIdOrMeasure === "string" ? localIdOrMeasure : measureLocalId(localIdOrMeasure);
        const indexes = this.dataAccess.findDataSeriesIndexes(localId);

        if (!indexes.length) {
            throw new Error(`there are no data series for measure with local id: ${localId}`);
        }

        return this.dataAccess.getDataSeries(indexes[0]);
    };

    public toArray = (): IDataSeries[] => {
        return Array.from(this);
    };
}

class DataSliceCollection implements IDataSliceCollection {
    public readonly count: number = 0;
    public readonly descriptors: Array<IAttributeDescriptor | ITotal> = [];

    constructor(private readonly dataAccess: DataAccessImpl) {
        const slicesDigest = this.dataAccess.getDataAccessPointers().slices;

        if (!slicesDigest) {
            return;
        }

        this.count = slicesDigest.count;
        this.descriptors = slicesDigest.descriptors;
    }

    public [Symbol.iterator] = (): Iterator<IDataSlice> => {
        return this.dataAccess.getDataSlicesIterator();
    };

    public toArray = (): IDataSlice[] => {
        return Array.from(this);
    };
}

class DataAccessMethods implements IDataAccessMethods {
    private readonly dataAccess: DataAccessImpl;
    private readonly seriesCollection: IDataSeriesCollection;
    private readonly slicesCollection: IDataSliceCollection;

    constructor(dataView: IDataView, config: DataAccessConfig) {
        this.dataAccess = new DataAccessImpl(dataView, config);
        this.seriesCollection = new DataSeriesCollection(this.dataAccess);
        this.slicesCollection = new DataSliceCollection(this.dataAccess);
    }

    public series(): IDataSeriesCollection {
        return this.seriesCollection;
    }

    public slices(): IDataSliceCollection {
        return this.slicesCollection;
    }
}

/**
 * @internal
 */
export function newDataAccessMethods(
    dataView: IDataView,
    config: DataAccessConfig = DefaultDataAccessConfig,
): IDataAccessMethods {
    return new DataAccessMethods(dataView, config);
}
