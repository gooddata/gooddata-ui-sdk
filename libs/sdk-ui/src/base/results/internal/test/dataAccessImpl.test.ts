// (C) 2020-2024 GoodData Corporation

import { ReferenceRecordings, ReferenceMd } from "@gooddata/reference-workspace";
import { DataViewFirstPage, dummyDataView, recordedDataView } from "@gooddata/sdk-backend-mockingbird";
import { newDataAccessMethods } from "../dataAccessMethods.js";
import { emptyDef, measureLocalId } from "@gooddata/sdk-model";
import { DataAccessConfig, DefaultDataAccessConfig } from "../../dataAccessConfig.js";
import { describe, it, expect } from "vitest";

describe("DataAccessMethods", () => {
    it("should handle empty data view", () => {
        const emptyDataView = dummyDataView(emptyDef("testWorkspace"));
        const dataAccess = newDataAccessMethods(emptyDataView);
        const series = dataAccess.series().toArray();
        const slices = dataAccess.slices().toArray();

        expect(series).toEqual([]);
        expect(slices).toEqual([]);
    });

    //
    //
    //

    describe("for data view with empty first dimension", () => {
        const DataViewWithEmptyFirstDim = recordedDataView(
            ReferenceRecordings.Scenarios.AreaChart.SingleMeasure,
        );

        it("should correctly extract data", () => {
            const dataAccess = newDataAccessMethods(DataViewWithEmptyFirstDim);
            const firstSeries = dataAccess.series().toArray()[0];

            const firstSeriesRawData = firstSeries.rawData();
            const firstSeriesDataPoints = Array.from(firstSeries);

            expect(firstSeriesDataPoints.map((dp) => dp.rawValue)).toEqual(firstSeriesRawData);
        });

        it("should return empty slices", () => {
            const dataAccess = newDataAccessMethods(DataViewWithEmptyFirstDim);
            const slices = dataAccess.slices().toArray();

            expect(slices).toEqual([]);
        });

        it("should correctly format data values", () => {
            const dataAccess = newDataAccessMethods(DataViewWithEmptyFirstDim);
            const firstSeries = dataAccess.series().toArray()[0];
            const firstSeriesDataPoints = Array.from(firstSeries);

            expect(firstSeriesDataPoints[0].rawValue).toEqual(116625456.54);
            expect(firstSeriesDataPoints[0].formattedValue()).toEqual("$116,625,456.54");
        });
    });

    //
    //
    //

    describe("for data view with two series and single-dim result", () => {
        const DataWithTwoSeriesAndNoSlices = recordedDataView(
            ReferenceRecordings.Scenarios.Headline.TwoMeasures,
        );

        it("should correctly extract data", () => {
            const dataAccess = newDataAccessMethods(DataWithTwoSeriesAndNoSlices);
            const series = dataAccess.series().toArray();

            expect(series.length).toEqual(2);

            for (const s of series) {
                const rawData = s.rawData();
                const dataPoints = Array.from(s).map((dp) => dp.rawValue);

                expect(rawData.length).toEqual(1);
                expect(dataPoints).toEqual(rawData);
            }
        });

        it("should return empty slices", () => {
            const dataAccess = newDataAccessMethods(DataWithTwoSeriesAndNoSlices);
            const slices = dataAccess.slices().toArray();

            expect(slices).toEqual([]);
        });

        it("should find series by measure", () => {
            const dataAccess = newDataAccessMethods(DataWithTwoSeriesAndNoSlices);

            expect(dataAccess.series().firstForMeasure(ReferenceMd.Amount)).toBeDefined();
            expect(dataAccess.series().firstForMeasure(ReferenceMd.Won)).toBeDefined();
        });

        it("should find all series by measure", () => {
            const dataAccess = newDataAccessMethods(DataWithTwoSeriesAndNoSlices);

            expect(Array.from(dataAccess.series().allForMeasure(ReferenceMd.Amount)).length).toEqual(1);
            expect(Array.from(dataAccess.series().allForMeasure(ReferenceMd.Won)).length).toEqual(1);
        });
    });

    //
    //
    //

    describe("for data with scopes series and empty slice dim", () => {
        const DataWithWithScopedSeriesAndNoSlices = recordedDataView(
            ReferenceRecordings.Scenarios.PivotTable.TwoMeasuresWithColumnAttribute,
            DataViewFirstPage,
        );

        it("should correctly extract scoped series", () => {
            const dataAccess = newDataAccessMethods(DataWithWithScopedSeriesAndNoSlices);
            const series = dataAccess.series().toArray();

            expect(series.length).toEqual(4);
        });

        it("should correctly extract data", () => {
            const dataAccess = newDataAccessMethods(DataWithWithScopedSeriesAndNoSlices);

            for (const s of dataAccess.series()) {
                const dataPoints = Array.from(s);
                const rawData = s.rawData();

                expect(dataPoints.length).toEqual(1);
                expect(dataPoints[0].sliceDesc).not.toBeDefined();
                expect(dataPoints.map((dp) => dp.rawValue)).toEqual(rawData);
            }
        });

        it("should return empty slices", () => {
            const dataAccess = newDataAccessMethods(DataWithWithScopedSeriesAndNoSlices);
            const slices = dataAccess.slices().toArray();

            expect(slices).toEqual([]);
        });
    });

    describe("for data with scoped series and slices", () => {
        const DataViewWithSeriesAndSlices = recordedDataView(
            ReferenceRecordings.Scenarios.PivotTable.TwoMeasuresWithRowAndColumnAttributes,
            DataViewFirstPage,
        );

        it("should correctly extract scoped series", () => {
            const dataAccess = newDataAccessMethods(DataViewWithSeriesAndSlices);
            const series = dataAccess.series().toArray();

            expect(series.length).toEqual(4);
        });

        it("should correctly extract slices", () => {
            const dataAccess = newDataAccessMethods(DataViewWithSeriesAndSlices);
            const slices = dataAccess.slices().toArray();

            expect(slices.length).toEqual(6);
        });

        it("should correctly extract data for series", () => {
            const dataAccess = newDataAccessMethods(DataViewWithSeriesAndSlices);

            for (const s of dataAccess.series()) {
                const dataPoints = Array.from(s);
                const rawData = s.rawData();

                expect(dataPoints.length).toEqual(6);
                expect(dataPoints.map((dp) => dp.rawValue)).toEqual(rawData);
                expect({
                    series: `${s.scopeTitles().join(" > ")} > ${s.measureTitle()}`,
                    slices: dataPoints.map((slice) => slice.sliceDesc!.sliceTitles()).join("|"),
                    rawData,
                }).toMatchSnapshot();
            }
        });
    });

    describe("for data with scoped series and slices with totals", () => {
        const DataViewWithSeriesAndSlices = recordedDataView(
            ReferenceRecordings.Scenarios.PivotTable.TwoMeasuresAndMultipleSubtotals,
            DataViewFirstPage,
        );

        it("should correctly extract scoped series", () => {
            const dataAccess = newDataAccessMethods(DataViewWithSeriesAndSlices);
            const series = dataAccess.series().toArray();

            expect(series.length).toEqual(8);
        });

        it("should correctly extract slices", () => {
            const dataAccess = newDataAccessMethods(DataViewWithSeriesAndSlices);
            const slices = dataAccess.slices().toArray();

            expect(slices.length).toEqual(24);
        });

        it("should correctly extract data for series", () => {
            const dataAccess = newDataAccessMethods(DataViewWithSeriesAndSlices);

            for (const s of dataAccess.series()) {
                const dataPoints = Array.from(s);
                const rawData = s.rawData();

                expect(dataPoints.length).toEqual(24);
                expect(dataPoints.map((dp) => dp.rawValue)).toEqual(rawData);
                expect({
                    series: `${s.scopeTitles().join(" > ")} > ${s.measureTitle()}`,
                    slices: dataPoints.map((slice) => slice.sliceDesc!.sliceTitles()).join("|"),
                    rawData,
                }).toMatchSnapshot();
            }
        });

        it("should translate headers", () => {
            const customConfig: DataAccessConfig = {
                ...DefaultDataAccessConfig,
                headerTranslator: (value: string | null) => `__${value}`,
            };

            const dataAccess = newDataAccessMethods(DataViewWithSeriesAndSlices, customConfig);
            const series = dataAccess.series().toArray();
            const slices = dataAccess.slices().toArray();

            const firstSeries = series[0];
            const firstSlice = slices[0];

            expect(firstSeries.scopeTitles()[0]).toMatch(/__.*/);
            expect(firstSlice.sliceTitles()[0]).toMatch(/__.*/);
        });

        it("should find series for measure", () => {
            const dataAccess = newDataAccessMethods(DataViewWithSeriesAndSlices);

            expect(dataAccess.series().firstForMeasure(ReferenceMd.Amount)).toBeDefined();
            expect(dataAccess.series().firstForMeasure(ReferenceMd.Won)).toBeDefined();
        });

        it("should find all series for measure", () => {
            const dataAccess = newDataAccessMethods(DataViewWithSeriesAndSlices);

            expect(Array.from(dataAccess.series().allForMeasure(ReferenceMd.Amount)).length).toEqual(4);

            for (const s of dataAccess.series().allForMeasure(ReferenceMd.Amount)) {
                expect(measureLocalId(s.descriptor.measureDefinition)).toEqual(
                    measureLocalId(ReferenceMd.Amount),
                );
            }

            expect(Array.from(dataAccess.series().allForMeasure(ReferenceMd.Won)).length).toEqual(4);

            for (const s of dataAccess.series().allForMeasure(ReferenceMd.Won)) {
                expect(measureLocalId(s.descriptor.measureDefinition)).toEqual(
                    measureLocalId(ReferenceMd.Won),
                );
            }
        });
    });
});
