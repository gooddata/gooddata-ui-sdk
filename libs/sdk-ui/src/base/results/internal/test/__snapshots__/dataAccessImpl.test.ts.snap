// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataAccessMethods > for data with scoped series and slices > should correctly extract data for series 1`] = `
{
  "rawData": [
    5099309.54,
    6614222.47,
    8269682.68,
    2761931.26,
    2664832.4,
    2607118.07,
  ],
  "series": "East Coast > Amount",
  "slices": "CompuSci|Educationly|Explorer|Grammar Plus|PhoenixSoft|WonderKid",
}
`;

exports[`DataAccessMethods > for data with scoped series and slices > should correctly extract data for series 2`] = `
{
  "rawData": [
    2115472,
    2307461.24,
    2749765.23,
    1401561.16,
    804564.3,
    1163935.72,
  ],
  "series": "East Coast > Won",
  "slices": "CompuSci|Educationly|Explorer|Grammar Plus|PhoenixSoft|WonderKid",
}
`;

exports[`DataAccessMethods > for data with scoped series and slices > should correctly extract data for series 3`] = `
{
  "rawData": [
    22123590.1,
    16332673,
    30326512.18,
    5280100.66,
    6861025.51,
    7684458.67,
  ],
  "series": "West Coast > Amount",
  "slices": "CompuSci|Educationly|Explorer|Grammar Plus|PhoenixSoft|WonderKid",
}
`;

exports[`DataAccessMethods > for data with scoped series and slices > should correctly extract data for series 4`] = `
{
  "rawData": [
    7122497.87,
    6550145.82,
    6497774.39,
    2571645.05,
    2569827.89,
    2456102.78,
  ],
  "series": "West Coast > Won",
  "slices": "CompuSci|Educationly|Explorer|Grammar Plus|PhoenixSoft|WonderKid",
}
`;

exports[`DataAccessMethods > for data with scoped series and slices with totals > should correctly extract data for series 1`] = `
{
  "rawData": [
    2174478.26,
    405767.14,
    2580245.4,
    1290122.7,
    2008366.85,
    590536.1,
    2598902.95,
    1299451.475,
    2610784.51,
    793939.81,
    3404724.32,
    1702362.16,
    1964931.93,
    268112.52,
    2233044.45,
    1116522.225,
    1827108.25,
    120416.22,
    1947524.47,
    973762.235,
    1374944.56,
    278916.58,
    1653861.1400000001,
    826930.5700000001,
  ],
  "series": "Exclude > East Coast > Amount",
  "slices": "CompuSci,Direct Sales|CompuSci,Inside Sales|CompuSci,sum|CompuSci,med|Educationly,Direct Sales|Educationly,Inside Sales|Educationly,sum|Educationly,med|Explorer,Direct Sales|Explorer,Inside Sales|Explorer,sum|Explorer,med|Grammar Plus,Direct Sales|Grammar Plus,Inside Sales|Grammar Plus,sum|Grammar Plus,med|PhoenixSoft,Direct Sales|PhoenixSoft,Inside Sales|PhoenixSoft,sum|PhoenixSoft,med|WonderKid,Direct Sales|WonderKid,Inside Sales|WonderKid,sum|WonderKid,med",
}
`;

exports[`DataAccessMethods > for data with scoped series and slices with totals > should correctly extract data for series 2`] = `
{
  "rawData": [
    883793.08,
    194123.92,
    null,
    538958.5,
    754830.5,
    283569.17,
    null,
    519199.83499999996,
    989426.6,
    430585.96,
    null,
    710006.28,
    826188.69,
    167291.74,
    null,
    496740.21499999997,
    229646.4,
    87001.62,
    null,
    158324.01,
    550627.31,
    119311.53,
    null,
    334969.42000000004,
  ],
  "series": "Exclude > East Coast > Won",
  "slices": "CompuSci,Direct Sales|CompuSci,Inside Sales|CompuSci,sum|CompuSci,med|Educationly,Direct Sales|Educationly,Inside Sales|Educationly,sum|Educationly,med|Explorer,Direct Sales|Explorer,Inside Sales|Explorer,sum|Explorer,med|Grammar Plus,Direct Sales|Grammar Plus,Inside Sales|Grammar Plus,sum|Grammar Plus,med|PhoenixSoft,Direct Sales|PhoenixSoft,Inside Sales|PhoenixSoft,sum|PhoenixSoft,med|WonderKid,Direct Sales|WonderKid,Inside Sales|WonderKid,sum|WonderKid,med",
}
`;

exports[`DataAccessMethods > for data with scoped series and slices with totals > should correctly extract data for series 3`] = `
{
  "rawData": [
    5917286.73,
    3156945.21,
    9074231.940000001,
    4537115.970000001,
    4088621.31,
    3053141.34,
    7141762.65,
    3570881.325,
    6011398.61,
    3135777.62,
    9147176.23,
    4573588.115,
    1060976.15,
    877547.53,
    1938523.68,
    969261.84,
    1980712.38,
    1358755.87,
    3339468.25,
    1669734.125,
    2542872.97,
    1330301.14,
    3873174.1100000003,
    1936587.0550000002,
  ],
  "series": "Exclude > West Coast > Amount",
  "slices": "CompuSci,Direct Sales|CompuSci,Inside Sales|CompuSci,sum|CompuSci,med|Educationly,Direct Sales|Educationly,Inside Sales|Educationly,sum|Educationly,med|Explorer,Direct Sales|Explorer,Inside Sales|Explorer,sum|Explorer,med|Grammar Plus,Direct Sales|Grammar Plus,Inside Sales|Grammar Plus,sum|Grammar Plus,med|PhoenixSoft,Direct Sales|PhoenixSoft,Inside Sales|PhoenixSoft,sum|PhoenixSoft,med|WonderKid,Direct Sales|WonderKid,Inside Sales|WonderKid,sum|WonderKid,med",
}
`;

exports[`DataAccessMethods > for data with scoped series and slices with totals > should correctly extract data for series 4`] = `
{
  "rawData": [
    2370306.08,
    1270431.27,
    null,
    1820368.675,
    1350035.81,
    1746096.06,
    null,
    1548065.935,
    2305188,
    1226965.38,
    null,
    1766076.69,
    389910.95,
    510727.21,
    null,
    450319.08,
    1150517.74,
    272152.06,
    null,
    711334.9,
    1109596.41,
    398431.88,
    null,
    754014.145,
  ],
  "series": "Exclude > West Coast > Won",
  "slices": "CompuSci,Direct Sales|CompuSci,Inside Sales|CompuSci,sum|CompuSci,med|Educationly,Direct Sales|Educationly,Inside Sales|Educationly,sum|Educationly,med|Explorer,Direct Sales|Explorer,Inside Sales|Explorer,sum|Explorer,med|Grammar Plus,Direct Sales|Grammar Plus,Inside Sales|Grammar Plus,sum|Grammar Plus,med|PhoenixSoft,Direct Sales|PhoenixSoft,Inside Sales|PhoenixSoft,sum|PhoenixSoft,med|WonderKid,Direct Sales|WonderKid,Inside Sales|WonderKid,sum|WonderKid,med",
}
`;

exports[`DataAccessMethods > for data with scoped series and slices with totals > should correctly extract data for series 5`] = `
{
  "rawData": [
    1989393.54,
    529670.6,
    2519064.14,
    1259532.07,
    3467615.16,
    547704.36,
    4015319.52,
    2007659.76,
    2935008.02,
    1929950.34,
    4864958.36,
    2432479.18,
    361332.16,
    167554.65,
    528886.8099999999,
    264443.40499999997,
    536526.06,
    180781.87,
    717307.93,
    358653.965,
    728206.16,
    225050.77,
    953256.93,
    476628.465,
  ],
  "series": "Include > East Coast > Amount",
  "slices": "CompuSci,Direct Sales|CompuSci,Inside Sales|CompuSci,sum|CompuSci,med|Educationly,Direct Sales|Educationly,Inside Sales|Educationly,sum|Educationly,med|Explorer,Direct Sales|Explorer,Inside Sales|Explorer,sum|Explorer,med|Grammar Plus,Direct Sales|Grammar Plus,Inside Sales|Grammar Plus,sum|Grammar Plus,med|PhoenixSoft,Direct Sales|PhoenixSoft,Inside Sales|PhoenixSoft,sum|PhoenixSoft,med|WonderKid,Direct Sales|WonderKid,Inside Sales|WonderKid,sum|WonderKid,med",
}
`;

exports[`DataAccessMethods > for data with scoped series and slices with totals > should correctly extract data for series 6`] = `
{
  "rawData": [
    817756.22,
    219798.78,
    null,
    518777.5,
    1160639.56,
    108422.01,
    null,
    634530.785,
    1106550.62,
    223202.05,
    null,
    664876.3350000001,
    259178.39,
    148902.34,
    null,
    204040.365,
    329035.25,
    158881.03,
    null,
    243958.14,
    476598.4,
    17398.48,
    null,
    246998.44,
  ],
  "series": "Include > East Coast > Won",
  "slices": "CompuSci,Direct Sales|CompuSci,Inside Sales|CompuSci,sum|CompuSci,med|Educationly,Direct Sales|Educationly,Inside Sales|Educationly,sum|Educationly,med|Explorer,Direct Sales|Explorer,Inside Sales|Explorer,sum|Explorer,med|Grammar Plus,Direct Sales|Grammar Plus,Inside Sales|Grammar Plus,sum|Grammar Plus,med|PhoenixSoft,Direct Sales|PhoenixSoft,Inside Sales|PhoenixSoft,sum|PhoenixSoft,med|WonderKid,Direct Sales|WonderKid,Inside Sales|WonderKid,sum|WonderKid,med",
}
`;

exports[`DataAccessMethods > for data with scoped series and slices with totals > should correctly extract data for series 7`] = `
{
  "rawData": [
    5501537.16,
    7547821,
    13049358.16,
    6524679.08,
    6623534.92,
    2567375.43,
    9190910.35,
    4595455.175,
    18472467,
    2706868.95,
    21179335.95,
    10589667.975,
    2476731.94,
    864845.04,
    3341576.98,
    1670788.49,
    1418895.61,
    2102661.65,
    3521557.26,
    1760778.63,
    2332594.72,
    1478689.84,
    3811284.5600000005,
    1905642.2800000003,
  ],
  "series": "Include > West Coast > Amount",
  "slices": "CompuSci,Direct Sales|CompuSci,Inside Sales|CompuSci,sum|CompuSci,med|Educationly,Direct Sales|Educationly,Inside Sales|Educationly,sum|Educationly,med|Explorer,Direct Sales|Explorer,Inside Sales|Explorer,sum|Explorer,med|Grammar Plus,Direct Sales|Grammar Plus,Inside Sales|Grammar Plus,sum|Grammar Plus,med|PhoenixSoft,Direct Sales|PhoenixSoft,Inside Sales|PhoenixSoft,sum|PhoenixSoft,med|WonderKid,Direct Sales|WonderKid,Inside Sales|WonderKid,sum|WonderKid,med",
}
`;

exports[`DataAccessMethods > for data with scoped series and slices with totals > should correctly extract data for series 8`] = `
{
  "rawData": [
    2039952.64,
    1441807.88,
    null,
    1740880.2599999998,
    2612041.49,
    841972.46,
    null,
    1727006.975,
    2097915.19,
    867705.82,
    null,
    1482810.505,
    1434946.69,
    236060.2,
    null,
    835503.445,
    583354.19,
    563803.9,
    null,
    573579.0449999999,
    676823.87,
    271250.62,
    null,
    474037.245,
  ],
  "series": "Include > West Coast > Won",
  "slices": "CompuSci,Direct Sales|CompuSci,Inside Sales|CompuSci,sum|CompuSci,med|Educationly,Direct Sales|Educationly,Inside Sales|Educationly,sum|Educationly,med|Explorer,Direct Sales|Explorer,Inside Sales|Explorer,sum|Explorer,med|Grammar Plus,Direct Sales|Grammar Plus,Inside Sales|Grammar Plus,sum|Grammar Plus,med|PhoenixSoft,Direct Sales|PhoenixSoft,Inside Sales|PhoenixSoft,sum|PhoenixSoft,med|WonderKid,Direct Sales|WonderKid,Inside Sales|WonderKid,sum|WonderKid,med",
}
`;
